src/BlazorWebView/src/Maui/Windows/WinUIWebViewManager.cs

```diff
@@ -71,52 +74,71 @@ public WinUIWebViewManager(
+			var url = eventArgs.Request.Uri;
 
-			// Get a deferral object so that WebView2 knows there's some async stuff going on. We call Complete() at the end of this method.
-			using var deferral = eventArgs.GetDeferral();
+			_logger.LogDebug("Intercepting request for {Url}.", url);
 
-			var requestUri = QueryStringHelper.RemovePossibleQueryString(eventArgs.Request.Uri);
-
-			_logger.HandlingWebRequest(requestUri);
-
-			var uri = new Uri(requestUri);
-			var relativePath = AppOriginUri.IsBaseOf(uri) ? AppOriginUri.MakeRelativeUri(uri).ToString() : null;
-
-			// Check if the uri is _framework/blazor.modules.json is a special case as the built-in file provider
-			// brings in a default implementation.
-			if (relativePath != null &&
-				string.Equals(relativePath, "_framework/blazor.modules.json", StringComparison.Ordinal) &&
-				await TryServeFromFolderAsync(eventArgs, allowFallbackOnHostPage: false, requestUri, relativePath))
-			{
-				_logger.ResponseContentBeingSent(requestUri, 200);
-			}
-			else if (TryGetResponseContent(requestUri, allowFallbackOnHostPage, out var statusCode, out var statusMessage, out var content, out var headers)
-				&& statusCode != 404)
+			// 1. First check if the app wants to modify or override the request.
+			if (WebRequestInterceptingWebView.TryInterceptResponseStream(_handler, _webview.CoreWebView2, eventArgs, url, _logger))
 			{
-				// First, call into WebViewManager to see if it has a framework file for this request. It will
-				// fall back to an IFileProvider, but on WinUI it's always a NullFileProvider, so that will never
-				// return a file.
-				var headerString = GetHeaderString(headers);
-				_logger.ResponseContentBeingSent(requestUri, statusCode);
-				eventArgs.Response = _coreWebView2Environment!.CreateWebResourceResponse(content.AsRandomAccessStream(), statusCode, statusMessage, headerString);
+				return;
 			}
-			else if (relativePath != null)
+
+			// 2. If this is an app request, then assume the request is for a Blazor resource.
+			var requestUri = QueryStringHelper.RemovePossibleQueryString(url);
+			if (new Uri(requestUri) is Uri uri)
 			{
-				await TryServeFromFolderAsync(
-					eventArgs,
-					allowFallbackOnHostPage,
-					requestUri,
-					relativePath);
+				// Unlike server-side code, we get told exactly why the browser is making the request,
+				// so we can be smarter about fallback. We can ensure that 'fetch' requests never result
+				// in fallback, for example.
+				var allowFallbackOnHostPage =
+					eventArgs.ResourceContext == CoreWebView2WebResourceContext.Document ||
+					eventArgs.ResourceContext == CoreWebView2WebResourceContext.Other; // e.g., dev tools requesting page source
+
+				// Get a deferral object so that WebView2 knows there's some async stuff going on. We call Complete() at the end of this method.
+				using var deferral = eventArgs.GetDeferral();
```

Comment:
if an exception happens before we call complete will the dispose be enough? 

Just making sure the deferral doesn't get stuck